### PR TITLE
Datadog propagator

### PR DESCRIPTION
**Pull request by kakkun61**
New propagator that is compliant with Datadog.

See: https://docs.datadoghq.com/tracing/other_telemetry/connect_logs_and_traces/opentelemetry/

### CODE REVIEW

File: api/src/OpenTelemetry/Trace/Id.hs [hunk 1]
```diff
@@ -44,7 +45,7 @@ module OpenTelemetry.Trace.Id (
   traceIdBaseEncodedText,
 
   -- * Working with 'SpanId's
-  SpanId,
```
Comment (iand675) [inline api/src/OpenTelemetry/Trace/Id.hs:47]: I don't really want to export `TraceId` / `SpanId` constructors for public consumption in general, but I'd be okay with exporting it from an `Internal` module.

File: propagators/datadog/hs-opentelemetry-propagator-datadog.cabal [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+  build-depends: bytestring,
+                 hs-opentelemetry-api,
+                 hs-opentelemetry-sdk,
+                 http-types,
+                 primitive,
+                 text
+  ghc-options: -Wcompat
+               -Wno-name-shadowing
+  if impl(ghc >= 6.4)
+    ghc-options: -Wincomplete-record-updates
```
Comment (iand675) [inline propagators/datadog/hs-opentelemetry-propagator-datadog.cabal:29]: Is this automatically generated from some tool?

**Review by iand675 — COMMENTED** [review]
This looks great! Thank you for submitting this. I have one small concern but otherwise will be happy to merge it once CI passes.

File: propagators/datadog/hs-opentelemetry-propagator-datadog.cabal (hunk 2 above)
Comment (kakkun61) [reply propagators/datadog/hs-opentelemetry-propagator-datadog.cabal:29]: This is from my project template https://github.com/kakkun61/haskell-project-template.

File: api/src/OpenTelemetry/Trace/Id.hs (hunk 1 above)
Comment (kakkun61) [reply api/src/OpenTelemetry/Trace/Id.hs]: I see.

How about this change? https://github.com/iand675/hs-opentelemetry/pull/49/commits/da6e4adf486139996f42761920eac42b16ccb59a

- move implementations to `OpenTelemetry.Internal.Trace.Id`
- control exposure at `OpenTelemetry.Trace.Id`

**Review by iand675 — APPROVED** [review]

**Comment by iand675** [comment]
Oh, one last thing– can you add the new propagator to the README.md?

**Comment by kakkun61** [comment]
Sure.

**Comment by kakkun61** [comment]
Hmm…:thinking:

```

Error: [S-7282]
       Stack failed to execute the build plan.
       
       While executing the build plan, Stack encountered the following errors:
       
       /home/runner/.stack/setup-exe-cache/x86_64-linux-nix/Cabal-simple_SvXsv1f__3.4.1.0_ghc-9.0.2: startProcess: exec: invalid argument (Bad file descriptor)
       
       /home/runner/.stack/setup-exe-cache/x86_64-linux-nix/Cabal-simple_SvXsv1f__3.4.1.0_ghc-9.0.2: startProcess: exec: invalid argument (Bad file descriptor)
```

https://github.com/iand675/hs-opentelemetry/actions/runs/3936967493/jobs/6733977106#step:6:330

**Review by iand675 — APPROVED** [review]

**Pull request merged**